File: Python solutions/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        encountered = {}
        left = 0
        longest = 0

        for right in range(len(s)):
            if s[right] in encountered.keys():
                left = max(left, encountered[s[right]] + 1)
            encountered[s[right]] = right
            longest = max(longest, right - left + 1)

        # print(longest)
        return longest

s1 = "abcabcbb"
ans1 = 3
assert Solution().lengthOfLongestSubstring(s1) == ans1, "Test failed!"

s2 = "bbbbb"
ans2 = 1
assert Solution().lengthOfLongestSubstring(s2) == ans2, "Test failed!"

s3 = "pwwkew"
ans3 = 3
assert Solution().lengthOfLongestSubstring(s3) == ans3, "Test failed!"

s4 = "au"
ans4 = 2
assert Solution().lengthOfLongestSubstring(s4) == ans4, "Test failed!"

s5 = "aab"
ans5 = 2
assert Solution().lengthOfLongestSubstring(s5) == ans5, "Test failed!"

print("\033[32mAll tests passed!\033[0m\n")

File: 3.Longest_Substring_Without_Repeating_Characters.cc
#include <iostream>
#include <cassert>
#include <unordered_set>
#include <unordered_map>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        std::unordered_map<char,int> encountered;
        int left = 0;
        int maxLength = 0;
        int n = s.size();

        for (int right = 0; right < n; ++right) {
            if (encountered.find(s[right]) != encountered.end()) {
                left = std::max(left, encountered[s[right]] + 1);
            } 
            encountered[s[right]] = right;
            maxLength = std::max(maxLength, right-left+1);
            
        }
        return maxLength;
    }
};


/*
    Iterate through s, check substrings with iteration using a sliding window
    a   b   c   a   b   c   b   b
    a
    a   b   
    a   b   c   a       -- repeat found, retract the window
        b   c   a   
        b   c   a   b   -- repeat found, retract the window
            c   a   b
            c   a   b   c -- repeat found, retract
                a   b   c
                a   b   c   b   -- repeat found, retract
                        c   b
                        c   b   b   -- repeat found
                                b

    For each letter add to the set
        if a repeat is found, compare length with existing maxLength and reset count and set
*/

int main() {
    Solution solution;

    assert(solution);

    std::cout << "\033[32mAll tests passed!\033[0m\n";
    return 0;
}